### PR TITLE
feat(amf): handling IMSI de-concealing failure cases

### PR DESCRIPTION
**Pull request by moinuddin1980**
Signed-off-by: Moinuddin Khan <moinuddin.khan@wavelabs.ai>

## Summary

Fixed the following:

1. During SUCI based registration the SUCI decryption is done to extract the MSIN. As, there is no validation for input UE public key, system generates Traceback in case of failure in MSIN extraction. Hence, added checks return proper error in case of invalid input parameter.
2. The logic of IMSI preparation using the decrypted MSIN, MCC and MNC was improper.
3.  Logic to extract ciphertext from the 5GS mobile identity from Scheme Output should consider variable length for ciphertext value.
4. The SUCI extension feature is used to decrypt the MSIN using the public key, msin and ciphertext. But, the APIs used with a naming convention of IMSI which was creating confusion. Hence, modified the APIs to include msin in naming convention instead of imsi.
5. SUCI-Profiles configured through Swagger-API are getting updated into the subscriber-db properly. 

## Test Plan

Tested the changes using Spirent tool to validate that no Traceback are generated in case of invalid UE public Key.
Ref Logs:
MME LOGS:
000639 Tue Apr 26 13:33:24 2022 7F0CEBDEF700 INFO  AMF-AP tasks/amf/Registration.cpp      :1080    Sending msg(grpc) to :[subscriberdb] for ue: [3] decrypt-imsi
000640 Tue Apr 26 13:33:24 2022 7F0CE25DC700 ERROR AMF-AP lib/n11/M5GSUCIRegistrationServi:0062    get_decrypt_imsi_info fails with code 5, Message : **Deconcealing IMSI failed**
000641 Tue Apr 26 13:33:24 2022 7F0CE25DC700 ERROR AMF-AP lib/n11/M5GSUCIRegistrationServi:0065    Deconcealing IMSI Failed, sending Registration Reject
[SUPI_MME3.log](https://github.com/magma/magma/files/8571018/SUPI_MME3.log)
[SUPI_SYS3.log](https://github.com/magma/magma/files/8571020/SUPI_SYS3.log)

Following SUCI related UT cases were modified and tested.
TestRegistrationProcSUCIExt, test_amf_ue_register_req_msg, test_amf_ue_suci_ext_register_req_msg etc.

## Additional Information
#11076 
#12750 
#12822

### CODE REVIEW

File: lte/gateway/python/magma/subscriberdb/protocols/m5g_auth_servicer.py [hunk 1]
```diff
@@ -176,14 +176,29 @@ def M5GDecryptImsiSUCIRegistration(self, request, context):
...
             )
 
-            msin_recv = home_network_info.unprotect(
-                request.ue_pubkey, request.ue_ciphertext,
-                request.ue_encrypted_mac,
-            )
+            if len(request.ue_pubkey) != 32:
+                set_grpc_err(
+                    context,
+                    StatusCode.NOT_FOUND,
```
Comment (sebathomas) [inline lte/gateway/python/magma/subscriberdb/protocols/m5g_auth_servicer.py]: Seems more like an INVALID_ARGUMENT to me?

File: lte/gateway/python/magma/subscriberdb/protocols/m5g_auth_servicer.py [hunk 2]
```diff
@@ -176,14 +176,29 @@ def M5GDecryptImsiSUCIRegistration(self, request, context):
...
-            aia.ue_msin_recv = msin_recv[:10]
-            logging.info("Deconcealed IMSI: %s", aia.ue_msin_recv)
-            return aia
+                if msin_recv != None:
+                    aia.ue_msin_recv = msin_recv[:10]
+                    logging.info("Deconcealed IMSI: %s", aia.ue_msin_recv)
+                else:
+                    set_grpc_err(
+                        context,
+                        StatusCode.NOT_FOUND,
```
Comment (sebathomas) [inline lte/gateway/python/magma/subscriberdb/protocols/m5g_auth_servicer.py]: Same here.

File: lte/gateway/python/magma/subscriberdb/protocols/m5g_auth_servicer.py (hunk 1 above)
Comment (moinuddin1980) [reply lte/gateway/python/magma/subscriberdb/protocols/m5g_auth_servicer.py]: Done

File: lte/gateway/python/magma/subscriberdb/protocols/m5g_auth_servicer.py (hunk 2 above)
Comment (moinuddin1980) [reply lte/gateway/python/magma/subscriberdb/protocols/m5g_auth_servicer.py]: Done

**Comment by panyogesh** [comment]
Signed-off-by: Moinuddin Khan [moinuddin.khan@wavelabs.ai](mailto:moinuddin.khan@wavelabs.ai)
 
## Summary
Fixed the following:
1. During SUCI based registration the SUCI decryption is done to extract the MSIN. 
As, there is no validation for input UE public key, system generates Traceback in case of failure in IMSI extraction. Hence, added checks return proper error in case of invalid input parameter.
2. The logic of IMSI preparation using the decrypted MSIN, MCC and MNC was improper.
3. The SUCI extension feature is used to decrypt the MSIN using the public key, msin and ciphertext. But, the APIs used with a naming convention of IMSI which was creating confusion. Hence, modified the APIs to include msin in naming convention instead of imsi.   
## Test Plan
Tested the changes using Spirent tool to validate that no Traceback are generated in case of invalid UE public Key. Ref Logs: 
MME LOGS: 
000639 Tue Apr 26 13:33:24 2022 7F0CEBDEF700 INFO AMF-AP tasks/amf/Registration.cpp :1080 Sending msg(grpc) to :[subscriberdb] for ue: [3] decrypt-imsi 
000640 Tue Apr 26 13:33:24 2022 7F0CE25DC700 ERROR AMF-AP lib/n11/M5GSUCIRegistrationServi:0062 get_decrypt_imsi_info fails with code 5, Message : **Deconcealing IMSI failed** 
000641 Tue Apr 26 13:33:24 2022 7F0CE25DC700 ERROR AMF-AP lib/n11/M5GSUCIRegistrationServi:0065 De-concealing IMSI Failed, sending Registration Reject
 [SUPI_MME3.log](https://github.com/magma/magma/files/8571018/SUPI_MME3.log) 
 [SUPI_SYS3.log](https://github.com/magma/magma/files/8571020/SUPI_SYS3.log)

Following SUCI related UT cases  were modified and tested.
- TestRegistrationProcSUCIExt, test_amf_ue_register_req_msg, test_amf_ue_suci_ext_register_req_msg etc.
 
## Additional Information
 #11076

Can you please update the summary. And testcase mapping to changes.

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp [hunk 3]
```diff
@@ -81,16 +81,18 @@ class AMFAppProcedureTest : public ::testing::Test {
...
-
-  const uint8_t intital_ue_message_suci_ext_hexbuf[65] = {
-      0x7e, 0x00, 0x41, 0x79, 0x00, 0x35, 0x01, 0x09, 0xf1, 0x07, 0x00,
-      0x00, 0x01, 0x04, 0xc8, 0xfc, 0x0c, 0xe5, 0x47, 0x9a, 0x51, 0x5d,
-      0xab, 0xf2, 0xf3, 0x45, 0xae, 0xb4, 0x66, 0x92, 0xd6, 0xff, 0x7a,
-      0x5f, 0x4f, 0x57, 0x2a, 0x47, 0x99, 0xf2, 0x33, 0x69, 0x35, 0x16,
-      0x40, 0x31, 0xbd, 0x3f, 0x84, 0x41, 0x26, 0xdf, 0x5b, 0x47, 0x06,
-      0x41, 0xe2, 0xa9, 0x57, 0x2e, 0x04, 0xf0, 0xf0, 0xf0, 0xf0};
+  itti_amf_decrypted_msin_info_ans_t decrypted_msin = {
+      .msin = {'9','7','6','5','4','5','6','6','0'}, .msin_length = 10, .result = 1, .ue_id = 1 };
```
[the remark this answers is not in the document]
Comment (moinuddin1980) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp]: Done

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GSMobileIdentity.cpp [hunk 4]
```diff
@@ -190,9 +200,9 @@ int M5GSMobileIdentityMsg::DecodeImsiMobileIdentityMsg(
           '\0';
     }
 
-    memcpy(&imsi->ciphertext, buffer + decoded, CIPHERTEXT_LENGTH);
-    decoded += CIPHERTEXT_LENGTH;
-    imsi->ciphertext[CIPHERTEXT_LENGTH] = '\0';
+    memcpy(&imsi->ciphertext, buffer + decoded, cipherTextLen);
```
Comment (panyogesh) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GSMobileIdentity.cpp]: Is it possible to use bstring. Kindly check M5GProtocolConfigurationOptions.cpp for the conversion.

Comment (moinuddin1980) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GSMobileIdentity.cpp]: Done

File: lte/gateway/python/magma/subscriberdb/client.py [hunk 5]
```diff
@@ -253,47 +253,52 @@ async def _get_all_subscribers(self) -> Optional[CloudSubscribersInfo]:
...
-                            home_network_public_key_identifier=int(k),
-                            protection_scheme=v.protection_scheme,
-                            home_network_public_key=v.home_network_public_key,
-                            home_network_private_key=v.home_network_private_key,
-                        ),
-                    )
-                logging.info("List of suciprofiles: %s", suciprofiles)
-                res = SuciProfileList(suci_profiles=suciprofiles)
-                return res
+            logging.info("List of suciprofiles: %s", suciprofiles)
```
Comment (panyogesh) [inline lte/gateway/python/magma/subscriberdb/client.py]: Remove the log or put it in debug mode.

File: lte/gateway/python/magma/subscriberdb/client.py [hunk 6]
```diff
@@ -253,49 +253,46 @@ async def _get_all_subscribers(self) -> Optional[CloudSubscribersInfo]:
...
-                res = await grpc_async_wrapper(
-                    subscriberdb_cloud_client.ListSuciProfiles.future(
-                        self.SUBSCRIBERDB_REQUEST_TIMEOUT,
+        suciprofiles = []
+        try:
+            req = ListSubscribersRequest(
+                page_size=self._subscriber_page_size,
+                page_token=req_page_token,
+            )
+            res = await grpc_async_wrapper(
```
Comment (panyogesh) [inline lte/gateway/python/magma/subscriberdb/client.py:264]: Seems ListSuciProfiles rpc does not takes any argument in. Is it required to pass the req parameter.

File: lte/gateway/python/magma/subscriberdb/client.py (hunk 5 above)
Comment (moinuddin1980) [reply lte/gateway/python/magma/subscriberdb/client.py]: Removed

File: lte/gateway/python/magma/subscriberdb/client.py (hunk 6 above)
Comment (moinuddin1980) [reply lte/gateway/python/magma/subscriberdb/client.py:264]: MicrosoftTeams-image (14)
It is required to pass the req parameter without which Traceback is generated.

File: lte/gateway/c/core/oai/include/amf_app_messages_types.h [hunk 7]
```diff
@@ -113,19 +113,19 @@ typedef struct itti_amf_subs_auth_info_ans_s {
 
 } itti_amf_subs_auth_info_ans_t;
 
-typedef struct itti_amf_decrypted_imsi_info_ans_s {
-  /* IMSI of the subscriber */
-  char imsi[IMSI_BCD_DIGITS_MAX + 1];
+typedef struct itti_amf_decrypted_msin_info_ans_s {
+  /* MSIN of the subscriber */
+  char msin[11];
```
Comment (ganeshg87) [inline lte/gateway/c/core/oai/include/amf_app_messages_types.h]: replace the hardcoded value with macro

**Review by ganeshg87 — APPROVED** [review]
lgtm...minor comment

File: lte/gateway/c/core/oai/include/amf_app_messages_types.h (hunk 7 above)
Comment (moinuddin1980) [reply lte/gateway/c/core/oai/include/amf_app_messages_types.h]: Done

**Comment by panyogesh** [comment]
LGTM

**Review by panyogesh — APPROVED** [review]

**Review by ulaskozat — APPROVED** [review]
LGTM

**Pull request merged**